iceoryx_posh/include/iceoryx_posh/popo/publisher.hpp

```diff
@@ -31,22 +31,31 @@ namespace popo
 /// This allows any publisher specialization to be stored as a reference by the Sample class.
 /// It is also needed to avoid circular dependencies between Sample and Publisher.
 ///
-template <typename T>
+template <typename T, typename H>
 class PublisherInterface
 {
   public:
-    virtual void publish(Sample<T>&& sample) noexcept = 0;
+    virtual void publish(Sample<T, H>&& sample) noexcept = 0;
 
     virtual ~PublisherInterface(){};
 
   protected:
     PublisherInterface() = default;
 };
-template <typename T, typename base_publisher_t = BasePublisher<>>
-class Publisher : public base_publisher_t, public PublisherInterface<T>
+template <typename T, typename H = mepoo::NoCustomHeader, typename base_publisher_t = BasePublisher<>>
+class Publisher : public base_publisher_t, public PublisherInterface<T, H>
 {
     static_assert(!std::is_void<T>::value, "Type must not be void. Use the UntypedPublisher for void types.");
-    static_assert(std::is_default_constructible<T>::value, "The Publisher requires default-constructable types.");
+    static_assert(!std::is_void<H>::value, "Custom header must not be void.");
+
+    /// @todo clarify in the PR if this can be removed for T.
+    /// This restriction is not necessary anymore, since `loan` takes ctor arguments.
+    static_assert(std::is_default_constructible<T>::value, "The Publisher requires default-constructable data types.");
+    static_assert(std::is_default_constructible<H>::value,
+                  "The Publisher requires default-constructable custom header types.");
+
+    static_assert(!std::is_const<T>::value, "The type must not be const.");
```

Comment:
Oh what about
1. when T or H is a reference?  (Here you send a sample with a reference to process local memory?!)
2. when T or H is a pointer? (Here you send a sample with a pointer to process local memory?!)

Could we get rid of the const check by using `std::remove_const<T>::type`. Then we can get rid of this `static_assert` and I think we can document this in the API and it wouldn't make the API more error prone ...

Reply:
I'm not sure about this. In the end it's a wrong usage since the data cannot be altered if it's const. We could remove the const, but it might be surprising even if we document it.

We probably should also catch pointer and references since this is also a clear misuse.

Reply:
Okay I am fine the const handling like it is now.
But two additional `static_assert` for references and pointers should be implemented here.